### ISSUE TITLE
在微调过程中，评估样例出现循环，到底是因为评估时没有惩罚参数，还是因为微调已经过拟合了？

### ISSUE THREAD

**Issue by zhanghui-china** [InternLM/xtuner#485]
150万菜谱 对 internlm2-chat-1.8b进行微调。

脚本如下：
···
# Copyright (c) OpenMMLab. All rights reserved.
import torch
from datasets import load_dataset
from mmengine.dataset import DefaultSampler
from mmengine.hooks import (CheckpointHook, DistSamplerSeedHook, IterTimerHook,
                            LoggerHook, ParamSchedulerHook)
from mmengine.optim import AmpOptimWrapper, CosineAnnealingLR, LinearLR
from peft import LoraConfig
from torch.optim import AdamW
from transformers import (AutoModelForCausalLM, AutoTokenizer,
                          BitsAndBytesConfig)

from xtuner.dataset import process_hf_dataset
from xtuner.dataset.collate_fns import default_collate_fn
from xtuner.dataset.map_fns import oasst1_map_fn, template_map_fn_factory
from xtuner.engine.hooks import (DatasetInfoHook, EvaluateChatHook,
                                 VarlenAttnArgsToMessageHubHook)
from xtuner.engine.runner import TrainLoop
from xtuner.model import SupervisedFinetune
from xtuner.utils import PROMPT_TEMPLATE

#######################################################################
#                          PART 1  Settings                           #
#######################################################################
# Model
pretrained_model_name_or_path = '/home/zhanghui/models/Shanghai_AI_Laboratory/internlm2-chat-1_8b'
use_varlen_attn = False

# Data
data_path = './dataset/tran_dataset_single_0.json'
prompt_template = PROMPT_TEMPLATE.internlm2_chat
max_length = 1024
pack_to_max_length = True

# Scheduler & Optimizer
batch_size = 12  # per_device
accumulative_counts = 8
dataloader_num_workers = 0
max_epochs = 1
optim_type = AdamW
#lr = 2e-4
lr = 1e-5
betas = (0.9, 0.999)
weight_decay = 0
max_norm = 1  # grad clip
warmup_ratio = 0.03

# Save
save_steps = 5039
save_total_limit = 10000  # Maximum checkpoints to keep (-1 means unlimited)

# Evaluate the generation performance during the training
evaluation_freq = 500
SYSTEM = ''
evaluation_inputs = [
    '酸菜鱼怎么做', '过桥米线怎么做','青椒土豆丝怎么做','小笼包怎么做','大煮干丝怎么做'
]

#######################################################################
#                      PART 2  Model & Tokenizer                      #
#######################################################################
tokenizer = dict(
    type=AutoTokenizer.from_pretrained,
    pretrained_model_name_or_path=pretrained_model_name_or_path,
    trust_remote_code=True,
    padding_side='right')

model = dict(
    type=SupervisedFinetune,
    use_varlen_attn=use_varlen_attn,
    llm=dict(
        type=AutoModelForCausalLM.from_pretrained,
        pretrained_model_name_or_path=pretrained_model_name_or_path,
        trust_remote_code=True,
        torch_dtype=torch.float16,
        quantization_config=dict(
            type=BitsAndBytesConfig,
            load_in_4bit=True,
            load_in_8bit=False,
            llm_int8_threshold=6.0,
            llm_int8_has_fp16_weight=False,
            bnb_4bit_compute_dtype=torch.float16,
            bnb_4bit_use_double_quant=True,
            bnb_4bit_quant_type='nf4')),
    lora=dict(
        type=LoraConfig,
        r=64,
        lora_alpha=16,
        lora_dropout=0.1,
        bias='none',
        task_type='CAUSAL_LM'))

#######################################################################
#                      PART 3  Dataset & Dataloader                   #
#######################################################################
train_dataset = dict(
    type=process_hf_dataset,
    #dataset=dict(type=load_dataset, path=data_path),
    dataset=dict(type=load_dataset, path='json', data_files=dict(train=data_path)),
    tokenizer=tokenizer,
    max_length=max_length,
    #dataset_map_fn=oasst1_map_fn,
    dataset_map_fn=None,
    template_map_fn=dict(
        type=template_map_fn_factory, template=prompt_template),
    remove_unused_columns=True,
    shuffle_before_pack=True,
    pack_to_max_length=pack_to_max_length,
    use_varlen_attn=use_varlen_attn)

train_dataloader = dict(
    batch_size=batch_size,
    num_workers=dataloader_num_workers,
    dataset=train_dataset,
    sampler=dict(type=DefaultSampler, shuffle=True),
    collate_fn=dict(type=default_collate_fn, use_varlen_attn=use_varlen_attn))

#######################################################################
#                    PART 4  Scheduler & Optimizer                    #
#######################################################################
# optimizer
optim_wrapper = dict(
    type=AmpOptimWrapper,
    optimizer=dict(
        type=optim_type, lr=lr, betas=betas, weight_decay=weight_decay),
    clip_grad=dict(max_norm=max_norm, error_if_nonfinite=False),
    accumulative_counts=accumulative_counts,
    loss_scale='dynamic',
    dtype='float16')

# learning policy
# More information: https://github.com/open-mmlab/mmengine/blob/main/docs/en/tutorials/param_scheduler.md  # noqa: E501
param_scheduler = [
    dict(
        type=LinearLR,
        start_factor=1e-5,
        by_epoch=True,
        begin=0,
        end=warmup_ratio * max_epochs,
        convert_to_iter_based=True),
    dict(
        type=CosineAnnealingLR,
        eta_min=0.0,
        by_epoch=True,
        begin=warmup_ratio * max_epochs,
        end=max_epochs,
        convert_to_iter_based=True)
]

# train, val, test setting
train_cfg = dict(type=TrainLoop, max_epochs=max_epochs)

#######################################################################
#                           PART 5  Runtime                           #
#######################################################################
# Log the dialogue periodically during the training process, optional
custom_hooks = [
    dict(type=DatasetInfoHook, tokenizer=tokenizer),
    dict(
        type=EvaluateChatHook,
        tokenizer=tokenizer,
        every_n_iters=evaluation_freq,
        evaluation_inputs=evaluation_inputs,
        system=SYSTEM,
        prompt_template=prompt_template)
]

if use_varlen_attn:
    custom_hooks += [dict(type=VarlenAttnArgsToMessageHubHook)]

# configure default hooks
default_hooks = dict(
    # record the time of every iteration.
    timer=dict(type=IterTimerHook),
    # print log every 10 iterations.
    logger=dict(type=LoggerHook, log_metric_by_epoch=False, interval=10),
    # enable the parameter scheduler.
    param_scheduler=dict(type=ParamSchedulerHook),
    # save checkpoint per `save_steps`.
    checkpoint=dict(
        type=CheckpointHook,
        by_epoch=False,
        interval=save_steps,
        max_keep_ckpts=save_total_limit),
    # set sampler seed in distributed evrionment.
    sampler_seed=dict(type=DistSamplerSeedHook),
)

# configure environment
env_cfg = dict(
    # whether to enable cudnn benchmark
    cudnn_benchmark=False,
    # set multi process parameters
    mp_cfg=dict(mp_start_method='fork', opencv_num_threads=0),
    # set distributed parameters
    dist_cfg=dict(backend='nccl'),
)

# set visualizer
visualizer = None

# set log level
log_level = 'INFO'

# load from which checkpoint
load_from = None

# whether to resume training from the loaded checkpoint
resume = False

# Defaults to use random seed and disable `deterministic`
randomness = dict(seed=None, deterministic=False)

# set log processor
log_processor = dict(by_epoch=False)

···
nvidia-smi：
<img width="585" alt="5cbfe429ab3f932ece585610d609146" src="https://github.com/InternLM/xtuner/assets/63148804/f100ef9f-c479-48e2-ad65-bb531fb967d2">

前500 iter：
<img width="1277" alt="4a3eb0034fbad986b6a4c35dd60817e" src="https://github.com/InternLM/xtuner/assets/63148804/dffad2e8-98a5-4eac-b8e7-3fe557d7c798">

501-100 iter：
<img width="1257" alt="3082fe3e8a29188f22a55a3082a0220" src="https://github.com/InternLM/xtuner/assets/63148804/30952091-bfe6-43ad-b137-c24b88268be0">

好像没多久就出现这种循环的情况了。

到了6000左右：
<img width="1280" alt="a0ed5fffe6ee7d9b7018aa9a2a9ede2" src="https://github.com/InternLM/xtuner/assets/63148804/95982894-b473-49cf-8db7-68715a35b481">
但是我感觉当前模型如果调整了惩罚参数，是仍然可以正常返回的。

也就是说，我们微调过程中，阶段性评估的时候，是不是可以送惩罚参数进去评估？

**Comment by zhanghui-china** [comment]
实际上xtuner chat的时候，加上惩罚参数，还是能正常返回的：
<img width="980" alt="c36f4de9f9d8c0cde95a0924678cce7" src="https://github.com/InternLM/xtuner/assets/63148804/73233ea1-cc43-48c0-a8c6-814986af99ed">

**Comment by LZHgrla** [comment]
@zhanghui-china Hi
循环输出在较“规则化”的文本输出时挺常见，增加惩罚参数可以解决这一问题。
像您这种情况，是由于在EvaluateChatHook时没有设置 repetition penalty（默认值为1），导致出现了循环。

**Comment by zhanghui-china** [comment]
> @zhanghui-china Hi 循环输出在较“规则化”的文本输出时挺常见，增加惩罚参数可以解决这一问题。 像您这种情况，是由于在EvaluateChatHook时没有设置 repetition penalty（默认值为1），导致出现了循环。

我明白这个原因。

但是我的看法是：在微调的过程中，每隔指定的iter数，就会打印评估问题及其答案，现在出现了循环的情况，这个评估就无法看出到这个步骤的是好还是坏。我们本来生成了这些iter的pth目录的时候，就希望能找到最佳的 微调模型时点。那么，是不是在每轮iter数评估的时候，就带着 repetition penalty进去评估？
如果可以，该怎么设置？
甚至，由于repetition penalty的数值可能需要调整，我能否设置多种repetition penalty对同一轮iter的pth进行 多角度的评估，以找到最佳的repetition penalty？

**Comment by LZHgrla** [comment]
https://github.com/InternLM/xtuner/pull/501
这个 PR 允许给 `EvaluateChatHook` 传参 `generation_kwargs` 来修改默认的对话设置。可以修改config实现参数的自定义设置，比如

```diff
custom_hooks = [
    dict(type=DatasetInfoHook, tokenizer=tokenizer),
    dict(
        type=EvaluateChatHook,
        tokenizer=tokenizer,
+       generation_kwargs={'max_new_tokens': 123, 'xxx': xxx},
        every_n_iters=evaluation_freq,
        evaluation_inputs=evaluation_inputs,
        system=SYSTEM,
        prompt_template=prompt_template)
]
```

如果需要每次对话都使用不同的对话配置，可以自行拓展一下 `EvaluateChatHook`，允许多个 generation_config 的选择
